File: examples/array.cpp
#include "example_common.h"
#include "iroha/iroha.h"

using namespace iroha;

IDesign *build_design() {
  IDesign *design = new IDesign;

  IModule *module = new IModule(design, "M");
  design->modules_.push_back(module);
  ITable *table = new ITable(module);
  module->tables_.push_back(table);
  IState *st1 = new IState(table);
  table->states_.push_back(st1);
  table->SetInitialState(st1);
  IState *st2 = new IState(table);
  table->states_.push_back(st2);
  DesignTool::AddNextState(st1, st2);

  IResource *array =
    DesignTool::CreateArrayResource(table, 10, 32,
				    /* is_external */ false, /* is_ram */ true);
  IRegister *addr = DesignTool::AllocConstNum(table, 10, 0);
  IRegister *write_data = DesignTool::AllocConstNum(table, 32, 123);
  IInsn *write_insn = new IInsn(array);
  write_insn->inputs_.push_back(addr);
  write_insn->inputs_.push_back(write_data);
  st1->insns_.push_back(write_insn);
  IInsn *read_insn = new IInsn(array);
  read_insn->inputs_.push_back(addr);
  IRegister *read_data = DesignTool::AllocRegister(table, "x", 32);
  read_insn->outputs_.push_back(read_data);
  st2->insns_.push_back(read_insn);

  DesignTool::Validate(design);
  return design;
}

File: examples/array.py
import sys
sys.path.append('../py')

from iroha import *
from iroha.iroha import *

d = IDesign()
mod = IModule(d, "mod")
tab = ITable(mod)
st1 = IState(tab)
st2 = IState(tab)
tab.initialSt = st1

tab.states.append(st1)
tab.states.append(st2)
design_tool.AddNextState(st1, st2)

array = design_tool.CreateArrayResource(
    tab, 10, 32,
    False, # is_external
    True # is_ram
)

write_insn = IInsn(array)
addr = design_tool.AllocConstNum(tab, False, 10, 0)
wdata = design_tool.AllocConstNum(tab, False, 32, 123)
write_insn.inputs.append(addr)
write_insn.inputs.append(wdata)
st1.insns.append(write_insn)

read_insn = IInsn(array)
read_insn.inputs.append(addr)
read_data = IRegister(tab, "r1")
read_insn.outputs.append(addr)
st2.insns.append(read_insn)
design_tool.ValidateIds(d)

w = DesignWriter(d)
w.Write()
